\documentclass[11pt]{article}
\usepackage{sectsty}
\allsectionsfont{\color{blue}\fontfamily{lmss}\selectfont}
\usepackage{fontspec}
\setmainfont{XCharter}

\usepackage{listings}
\lstset{
basicstyle=\small\ttfamily,
tabsize=8,
columns=flexible,
breaklines=true,
frame=tb,
rulecolor=\color[rgb]{0.8,0.8,0.7},
backgroundcolor=\color[rgb]{1,1,0.91},
postbreak=\raisebox{0ex}[0ex][0ex]{\ensuremath{\color{red}\hookrightarrow\space}}
}
\usepackage{fontawesome}


\usepackage{mdframed}
\newmdenv[
  backgroundcolor=gray,
  fontcolor=white,
  nobreak=true,
]{terminalinput}



\usepackage{parskip}


    \usepackage[breakable]{tcolorbox}
    \usepackage{parskip} % Stop auto-indenting (to mimic markdown behaviour)


    % Basic figure setup, for now with no caption control since it's done
    % automatically by Pandoc (which extracts ![](path) syntax from Markdown).
    \usepackage{graphicx}
    % Maintain compatibility with old templates. Remove in nbconvert 6.0
    \let\Oldincludegraphics\includegraphics
    % Ensure that by default, figures have no caption (until we provide a
    % proper Figure object with a Caption API and a way to capture that
    % in the conversion process - todo).
    \usepackage{caption}
    \DeclareCaptionFormat{nocaption}{}
    \captionsetup{format=nocaption,aboveskip=0pt,belowskip=0pt}

    \usepackage{float}
    \floatplacement{figure}{H} % forces figures to be placed at the correct location
    \usepackage{xcolor} % Allow colors to be defined
    \usepackage{enumerate} % Needed for markdown enumerations to work
    \usepackage{geometry} % Used to adjust the document margins
    \usepackage{amsmath} % Equations
    \usepackage{amssymb} % Equations
    \usepackage{textcomp} % defines textquotesingle
    % Hack from http://tex.stackexchange.com/a/47451/13684:
    \AtBeginDocument{%
        \def\PYZsq{\textquotesingle}% Upright quotes in Pygmentized code
    }
    \usepackage{upquote} % Upright quotes for verbatim code
    \usepackage{eurosym} % defines \euro

    \usepackage{iftex}
    \ifPDFTeX
        \usepackage[T1]{fontenc}
        \IfFileExists{alphabeta.sty}{
              \usepackage{alphabeta}
          }{
              \usepackage[mathletters]{ucs}
              \usepackage[utf8x]{inputenc}
          }
    \else
        \usepackage{fontspec}
        \usepackage{unicode-math}
    \fi

    \usepackage{fancyvrb} % verbatim replacement that allows latex
    \usepackage{grffile} % extends the file name processing of package graphics
                         % to support a larger range
    \makeatletter % fix for old versions of grffile with XeLaTeX
    \@ifpackagelater{grffile}{2019/11/01}
    {
      % Do nothing on new versions
    }
    {
      \def\Gread@@xetex#1{%
        \IfFileExists{"\Gin@base".bb}%
        {\Gread@eps{\Gin@base.bb}}%
        {\Gread@@xetex@aux#1}%
      }
    }
    \makeatother
    \usepackage[Export]{adjustbox} % Used to constrain images to a maximum size
    \adjustboxset{max size={0.9\linewidth}{0.9\paperheight}}

    % The hyperref package gives us a pdf with properly built
    % internal navigation ('pdf bookmarks' for the table of contents,
    % internal cross-reference links, web links for URLs, etc.)
    \usepackage{hyperref}
    % The default LaTeX title has an obnoxious amount of whitespace. By default,
    % titling removes some of it. It also provides customization options.
    \usepackage{titling}
    \usepackage{longtable} % longtable support required by pandoc >1.10
    \usepackage{booktabs}  % table support for pandoc > 1.12.2
    \usepackage{array}     % table support for pandoc >= 2.11.3
    \usepackage{calc}      % table minipage width calculation for pandoc >= 2.11.1
    \usepackage[inline]{enumitem} % IRkernel/repr support (it uses the enumerate* environment)
    \usepackage[normalem]{ulem} % ulem is needed to support strikethroughs (\sout)
                                % normalem makes italics be italics, not underlines
    \usepackage{mathrsfs}



    % Colors for the hyperref package
    \definecolor{urlcolor}{rgb}{0,.145,.698}
    \definecolor{linkcolor}{rgb}{.71,0.21,0.01}
    \definecolor{citecolor}{rgb}{.12,.54,.11}

    % ANSI colors
    \definecolor{ansi-black}{HTML}{3E424D}
    \definecolor{ansi-black-intense}{HTML}{282C36}
    \definecolor{ansi-red}{HTML}{E75C58}
    \definecolor{ansi-red-intense}{HTML}{B22B31}
    \definecolor{ansi-green}{HTML}{00A250}
    \definecolor{ansi-green-intense}{HTML}{007427}
    \definecolor{ansi-yellow}{HTML}{DDB62B}
    \definecolor{ansi-yellow-intense}{HTML}{B27D12}
    \definecolor{ansi-blue}{HTML}{208FFB}
    \definecolor{ansi-blue-intense}{HTML}{0065CA}
    \definecolor{ansi-magenta}{HTML}{D160C4}
    \definecolor{ansi-magenta-intense}{HTML}{A03196}
    \definecolor{ansi-cyan}{HTML}{60C6C8}
    \definecolor{ansi-cyan-intense}{HTML}{258F8F}
    \definecolor{ansi-white}{HTML}{C5C1B4}
    \definecolor{ansi-white-intense}{HTML}{A1A6B2}
    \definecolor{ansi-default-inverse-fg}{HTML}{FFFFFF}
    \definecolor{ansi-default-inverse-bg}{HTML}{000000}

    % common color for the border for error outputs.
    \definecolor{outerrorbackground}{HTML}{FFDFDF}

    % commands and environments needed by pandoc snippets
    % extracted from the output of `pandoc -s`
    \providecommand{\tightlist}{%
      \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
    \DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
    % Add ',fontsize=\small' for more characters per line
    \newenvironment{Shaded}{}{}
    \newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{\textbf{{#1}}}}
    \newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.56,0.13,0.00}{{#1}}}
    \newcommand{\DecValTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\FloatTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\CharTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\StringTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\CommentTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textit{{#1}}}}
    \newcommand{\OtherTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{{#1}}}
    \newcommand{\AlertTok}[1]{\textcolor[rgb]{1.00,0.00,0.00}{\textbf{{#1}}}}
    \newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.02,0.16,0.49}{{#1}}}
    \newcommand{\RegionMarkerTok}[1]{{#1}}
    \newcommand{\ErrorTok}[1]{\textcolor[rgb]{1.00,0.00,0.00}{\textbf{{#1}}}}
    \newcommand{\NormalTok}[1]{{#1}}

    % Additional commands for more recent versions of Pandoc
    \newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.53,0.00,0.00}{{#1}}}
    \newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.73,0.40,0.53}{{#1}}}
    \newcommand{\ImportTok}[1]{{#1}}
    \newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.73,0.13,0.13}{\textit{{#1}}}}
    \newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\VariableTok}[1]{\textcolor[rgb]{0.10,0.09,0.49}{{#1}}}
    \newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{\textbf{{#1}}}}
    \newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.40,0.40,0.40}{{#1}}}
    \newcommand{\BuiltInTok}[1]{{#1}}
    \newcommand{\ExtensionTok}[1]{{#1}}
    \newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.74,0.48,0.00}{{#1}}}
    \newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.49,0.56,0.16}{{#1}}}
    \newcommand{\InformationTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\WarningTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}


    % Define a nice break command that doesn't care if a line doesn't already
    % exist.
    \def\br{\hspace*{\fill} \\* }
    % Math Jax compatibility definitions
    \def\gt{>}
    \def\lt{<}
    \let\Oldtex\TeX
    \let\Oldlatex\LaTeX
    \renewcommand{\TeX}{\textrm{\Oldtex}}
    \renewcommand{\LaTeX}{\textrm{\Oldlatex}}
    % Document parameters
    % Document title
    \title{index}





% Pygments definitions
\makeatletter
\def\PY@reset{\let\PY@it=\relax \let\PY@bf=\relax%
    \let\PY@ul=\relax \let\PY@tc=\relax%
    \let\PY@bc=\relax \let\PY@ff=\relax}
\def\PY@tok#1{\csname PY@tok@#1\endcsname}
\def\PY@toks#1+{\ifx\relax#1\empty\else%
    \PY@tok{#1}\expandafter\PY@toks\fi}
\def\PY@do#1{\PY@bc{\PY@tc{\PY@ul{%
    \PY@it{\PY@bf{\PY@ff{#1}}}}}}}
\def\PY#1#2{\PY@reset\PY@toks#1+\relax+\PY@do{#2}}

\@namedef{PY@tok@w}{\def\PY@tc##1{\textcolor[rgb]{0.73,0.73,0.73}{##1}}}
\@namedef{PY@tok@c}{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.24,0.48,0.48}{##1}}}
\@namedef{PY@tok@cp}{\def\PY@tc##1{\textcolor[rgb]{0.61,0.40,0.00}{##1}}}
\@namedef{PY@tok@k}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\@namedef{PY@tok@kp}{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\@namedef{PY@tok@kt}{\def\PY@tc##1{\textcolor[rgb]{0.69,0.00,0.25}{##1}}}
\@namedef{PY@tok@o}{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\@namedef{PY@tok@ow}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.67,0.13,1.00}{##1}}}
\@namedef{PY@tok@nb}{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\@namedef{PY@tok@nf}{\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\@namedef{PY@tok@nc}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\@namedef{PY@tok@nn}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\@namedef{PY@tok@ne}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.80,0.25,0.22}{##1}}}
\@namedef{PY@tok@nv}{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\@namedef{PY@tok@no}{\def\PY@tc##1{\textcolor[rgb]{0.53,0.00,0.00}{##1}}}
\@namedef{PY@tok@nl}{\def\PY@tc##1{\textcolor[rgb]{0.46,0.46,0.00}{##1}}}
\@namedef{PY@tok@ni}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.44,0.44,0.44}{##1}}}
\@namedef{PY@tok@na}{\def\PY@tc##1{\textcolor[rgb]{0.41,0.47,0.13}{##1}}}
\@namedef{PY@tok@nt}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\@namedef{PY@tok@nd}{\def\PY@tc##1{\textcolor[rgb]{0.67,0.13,1.00}{##1}}}
\@namedef{PY@tok@s}{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\@namedef{PY@tok@sd}{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\@namedef{PY@tok@si}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.64,0.35,0.47}{##1}}}
\@namedef{PY@tok@se}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.67,0.36,0.12}{##1}}}
\@namedef{PY@tok@sr}{\def\PY@tc##1{\textcolor[rgb]{0.64,0.35,0.47}{##1}}}
\@namedef{PY@tok@ss}{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\@namedef{PY@tok@sx}{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\@namedef{PY@tok@m}{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\@namedef{PY@tok@gh}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\@namedef{PY@tok@gu}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.50,0.00,0.50}{##1}}}
\@namedef{PY@tok@gd}{\def\PY@tc##1{\textcolor[rgb]{0.63,0.00,0.00}{##1}}}
\@namedef{PY@tok@gi}{\def\PY@tc##1{\textcolor[rgb]{0.00,0.52,0.00}{##1}}}
\@namedef{PY@tok@gr}{\def\PY@tc##1{\textcolor[rgb]{0.89,0.00,0.00}{##1}}}
\@namedef{PY@tok@ge}{\let\PY@it=\textit}
\@namedef{PY@tok@gs}{\let\PY@bf=\textbf}
\@namedef{PY@tok@gp}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\@namedef{PY@tok@go}{\def\PY@tc##1{\textcolor[rgb]{0.44,0.44,0.44}{##1}}}
\@namedef{PY@tok@gt}{\def\PY@tc##1{\textcolor[rgb]{0.00,0.27,0.87}{##1}}}
\@namedef{PY@tok@err}{\def\PY@bc##1{{\setlength{\fboxsep}{\string -\fboxrule}\fcolorbox[rgb]{1.00,0.00,0.00}{1,1,1}{\strut ##1}}}}
\@namedef{PY@tok@kc}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\@namedef{PY@tok@kd}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\@namedef{PY@tok@kn}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\@namedef{PY@tok@kr}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\@namedef{PY@tok@bp}{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\@namedef{PY@tok@fm}{\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\@namedef{PY@tok@vc}{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\@namedef{PY@tok@vg}{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\@namedef{PY@tok@vi}{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\@namedef{PY@tok@vm}{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\@namedef{PY@tok@sa}{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\@namedef{PY@tok@sb}{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\@namedef{PY@tok@sc}{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\@namedef{PY@tok@dl}{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\@namedef{PY@tok@s2}{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\@namedef{PY@tok@sh}{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\@namedef{PY@tok@s1}{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\@namedef{PY@tok@mb}{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\@namedef{PY@tok@mf}{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\@namedef{PY@tok@mh}{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\@namedef{PY@tok@mi}{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\@namedef{PY@tok@il}{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\@namedef{PY@tok@mo}{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\@namedef{PY@tok@ch}{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.24,0.48,0.48}{##1}}}
\@namedef{PY@tok@cm}{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.24,0.48,0.48}{##1}}}
\@namedef{PY@tok@cpf}{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.24,0.48,0.48}{##1}}}
\@namedef{PY@tok@c1}{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.24,0.48,0.48}{##1}}}
\@namedef{PY@tok@cs}{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.24,0.48,0.48}{##1}}}

\def\PYZbs{\char`\\}
\def\PYZus{\char`\_}
\def\PYZob{\char`\{}
\def\PYZcb{\char`\}}
\def\PYZca{\char`\^}
\def\PYZam{\char`\&}
\def\PYZlt{\char`\<}
\def\PYZgt{\char`\>}
\def\PYZsh{\char`\#}
\def\PYZpc{\char`\%}
\def\PYZdl{\char`\$}
\def\PYZhy{\char`\-}
\def\PYZsq{\char`\'}
\def\PYZdq{\char`\"}
\def\PYZti{\char`\~}
% for compatibility with earlier versions
\def\PYZat{@}
\def\PYZlb{[}
\def\PYZrb{]}
\makeatother


    % For linebreaks inside Verbatim environment from package fancyvrb.
    \makeatletter
        \newbox\Wrappedcontinuationbox
        \newbox\Wrappedvisiblespacebox
        \newcommand*\Wrappedvisiblespace {\textcolor{red}{\textvisiblespace}}
        \newcommand*\Wrappedcontinuationsymbol {\textcolor{red}{\llap{\tiny$\m@th\hookrightarrow$}}}
        \newcommand*\Wrappedcontinuationindent {3ex }
        \newcommand*\Wrappedafterbreak {\kern\Wrappedcontinuationindent\copy\Wrappedcontinuationbox}
        % Take advantage of the already applied Pygments mark-up to insert
        % potential linebreaks for TeX processing.
        %        {, <, #, %, $, ' and ": go to next line.
        %        _, }, ^, &, >, - and ~: stay at end of broken line.
        % Use of \textquotesingle for straight quote.
        \newcommand*\Wrappedbreaksatspecials {%
            \def\PYGZus{\discretionary{\char`\_}{\Wrappedafterbreak}{\char`\_}}%
            \def\PYGZob{\discretionary{}{\Wrappedafterbreak\char`\{}{\char`\{}}%
            \def\PYGZcb{\discretionary{\char`\}}{\Wrappedafterbreak}{\char`\}}}%
            \def\PYGZca{\discretionary{\char`\^}{\Wrappedafterbreak}{\char`\^}}%
            \def\PYGZam{\discretionary{\char`\&}{\Wrappedafterbreak}{\char`\&}}%
            \def\PYGZlt{\discretionary{}{\Wrappedafterbreak\char`\<}{\char`\<}}%
            \def\PYGZgt{\discretionary{\char`\>}{\Wrappedafterbreak}{\char`\>}}%
            \def\PYGZsh{\discretionary{}{\Wrappedafterbreak\char`\#}{\char`\#}}%
            \def\PYGZpc{\discretionary{}{\Wrappedafterbreak\char`\%}{\char`\%}}%
            \def\PYGZdl{\discretionary{}{\Wrappedafterbreak\char`\$}{\char`\$}}%
            \def\PYGZhy{\discretionary{\char`\-}{\Wrappedafterbreak}{\char`\-}}%
            \def\PYGZsq{\discretionary{}{\Wrappedafterbreak\textquotesingle}{\textquotesingle}}%
            \def\PYGZdq{\discretionary{}{\Wrappedafterbreak\char`\"}{\char`\"}}%
            \def\PYGZti{\discretionary{\char`\~}{\Wrappedafterbreak}{\char`\~}}%
        }
        % Some characters . , ; ? ! / are not pygmentized.
        % This macro makes them "active" and they will insert potential linebreaks
        \newcommand*\Wrappedbreaksatpunct {%
            \lccode`\~`\.\lowercase{\def~}{\discretionary{\hbox{\char`\.}}{\Wrappedafterbreak}{\hbox{\char`\.}}}%
            \lccode`\~`\,\lowercase{\def~}{\discretionary{\hbox{\char`\,}}{\Wrappedafterbreak}{\hbox{\char`\,}}}%
            \lccode`\~`\;\lowercase{\def~}{\discretionary{\hbox{\char`\;}}{\Wrappedafterbreak}{\hbox{\char`\;}}}%
            \lccode`\~`\:\lowercase{\def~}{\discretionary{\hbox{\char`\:}}{\Wrappedafterbreak}{\hbox{\char`\:}}}%
            \lccode`\~`\?\lowercase{\def~}{\discretionary{\hbox{\char`\?}}{\Wrappedafterbreak}{\hbox{\char`\?}}}%
            \lccode`\~`\!\lowercase{\def~}{\discretionary{\hbox{\char`\!}}{\Wrappedafterbreak}{\hbox{\char`\!}}}%
            \lccode`\~`\/\lowercase{\def~}{\discretionary{\hbox{\char`\/}}{\Wrappedafterbreak}{\hbox{\char`\/}}}%
            \catcode`\.\active
            \catcode`\,\active
            \catcode`\;\active
            \catcode`\:\active
            \catcode`\?\active
            \catcode`\!\active
            \catcode`\/\active
            \lccode`\~`\~
        }
    \makeatother

    \let\OriginalVerbatim=\Verbatim
    \makeatletter
    \renewcommand{\Verbatim}[1][1]{%
        %\parskip\z@skip
        \sbox\Wrappedcontinuationbox {\Wrappedcontinuationsymbol}%
        \sbox\Wrappedvisiblespacebox {\FV@SetupFont\Wrappedvisiblespace}%
        \def\FancyVerbFormatLine ##1{\hsize\linewidth
            \vtop{\raggedright\hyphenpenalty\z@\exhyphenpenalty\z@
                \doublehyphendemerits\z@\finalhyphendemerits\z@
                \strut ##1\strut}%
        }%
        % If the linebreak is at a space, the latter will be displayed as visible
        % space at end of first line, and a continuation symbol starts next line.
        % Stretch/shrink are however usually zero for typewriter font.
        \def\FV@Space {%
            \nobreak\hskip\z@ plus\fontdimen3\font minus\fontdimen4\font
            \discretionary{\copy\Wrappedvisiblespacebox}{\Wrappedafterbreak}
            {\kern\fontdimen2\font}%
        }%

        % Allow breaks at special characters using \PYG... macros.
        \Wrappedbreaksatspecials
        % Breaks at punctuation characters . , ; ? ! and / need catcode=\active
        \OriginalVerbatim[#1,codes*=\Wrappedbreaksatpunct]%
    }
    \makeatother

    % Exact colors from NB
    \definecolor{incolor}{HTML}{303F9F}
    \definecolor{outcolor}{HTML}{D84315}
    \definecolor{cellborder}{HTML}{CFCFCF}
    \definecolor{cellbackground}{HTML}{F7F7F7}

    % prompt
    \makeatletter
    \newcommand{\boxspacing}{\kern\kvtcb@left@rule\kern\kvtcb@boxsep}
    \makeatother
    \newcommand{\prompt}[4]{
        {\ttfamily\llap{{\color{#2}[#3]:\hspace{3pt}#4}}\vspace{-\baselineskip}}
    }



    % Prevent overflowing lines due to hard-to-break entities
    \sloppy
    % Setup hyperref package
    \hypersetup{
      breaklinks=true,  % so long urls are correctly broken across lines
      colorlinks=true,
      urlcolor=urlcolor,
      linkcolor=linkcolor,
      citecolor=citecolor,
      }
    % Slightly bigger margins than the latex defaults

    \geometry{verbose,tmargin=1in,bmargin=1in,lmargin=1in,rmargin=1in}



\renewcommand{\PY}[2]{{#2}}
\usepackage{fancyhdr}
\pagestyle{fancy}
\rhead{\color{gray}\sf\small\rightmark}
\lhead{\nouppercase{\color{gray}\sf\small\leftmark}}
\cfoot{\color{gray}\sf\thepage}
\renewcommand{\footrulewidth}{1pt}
\begin{document}





    \hypertarget{ngs-data-formats-and-qc}{%
\section{NGS Data formats and QC}\label{ngs-data-formats-and-qc}}

\hypertarget{introduction}{%
\subsection{Introduction}\label{introduction}}

There are several file formats for storing Next Generation Sequencing
(NGS) data. In this tutorial we will look at some of the most common
formats for storing NGS reads and variant data. We will cover the
following formats:

\textbf{FASTQ} - This format stores unaligned read sequences with base
qualities\\
\textbf{SAM/BAM} - This format stores unaligned or aligned reads (text
and binary formats)\\
\textbf{CRAM} - This format is similar to BAM but has better compression
than BAM\\
\textbf{VCF/BCF} - Flexible variant call format for storing SNPs,
indels, structural variations (text and binary formats)

Further to understanding the different file formats, it is important to
remember that all sequencing platforms have technical limitations that
can introduce biases in your sequencing data. Because of this it is very
important to check the quality of the data before starting any analysis,
whether you are planning to use something you have sequenced yourself or
publicly available data. In the latter part of this tutorial we will
describe how to perform a QC assessment for your NGS data, and also
suggest how to identify possible contamination.

\hypertarget{learning-outcomes}{%
\subsection{Learning outcomes}\label{learning-outcomes}}

On completion of the tutorial, you can expect to be able to:

\begin{itemize}
\tightlist
\item
  Describe the different NGS data formats available (FASTQ, SAM/BAM,
  CRAM, VCF/BCF)
\item
  Perform a QC assessment of high throughput sequence data
\item
  Identify possible contamination in high throughput sequence data
\end{itemize}

\hypertarget{tutorial-sections}{%
\subsection{Tutorial sections}\label{tutorial-sections}}

This tutorial comprises the following sections:\\
1. \href{formats.ipynb}{Data formats}\\
2. \href{assessment.ipynb}{QC assessment}\\
3. \href{contamination.ipynb}{Identifying contamination}

\hypertarget{authors}{%
\subsection{Authors}\label{authors}}

This tutorial was written by \href{https://github.com/ssjunnebo}{Sara
Sjunnebo} and \href{https://github.com/jacquikeane}{Jacqui Keane} based
on material from \href{https://github.com/pd3}{Petr Danecek} and
\href{https://github.com/tk2}{Thomas Keane}.

\hypertarget{running-the-commands-from-this-tutorial}{%
\subsection{Running the commands from this
tutorial}\label{running-the-commands-from-this-tutorial}}

You can follow this tutorial by running all the commands provided on the
\texttt{bioinfsrv} server. To get started, connect to the
\texttt{bioinfsrv} server using the instructions provided earlier and
type all the commands you see into the terminal window. Remember, the
terminal window is similar to the ``Command Prompt'' window on MS
Windows systems, which allows the user to type DOS commands to manage
files.

Start by typing the command below:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n+nb}{cd}\PY{+w}{ }\PYZti{}/course\PYZus{}data/data\PYZus{}formats/data
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{lets-get-started}{%
\subsection{Let's get started!}\label{lets-get-started}}

This tutorial assumes that you have samtools, bcftools, fastqc and
bactinspector on your computer. The easiest way to do this is by
creating a Conda environment called \texttt{qc} and install the listed
software using conda. This has already been setup on the server for you.
To activate the environment use

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
conda\PY{+w}{ }activate\PY{+w}{ }qc
\end{Verbatim}
\end{tcolorbox}

    After your environment is activated you can run the following commands:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
samtools\PY{+w}{ }\PYZhy{}\PYZhy{}help
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bcftools\PY{+w}{ }\PYZhy{}\PYZhy{}help
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
fastqc\PY{+w}{ }\PYZhy{}\PYZhy{}help
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bactinspector\PY{+w}{ }\PYZhy{}h
\end{Verbatim}
\end{tcolorbox}

    This should return the help message for these tools.

To get started with the tutorial, head to the first section:
\href{formats.ipynb}{Data formats}


    % Add a bibliography block to the postdoc



\newpage





    \hypertarget{data-formats-for-ngs-data}{%
\section{Data formats for NGS data}\label{data-formats-for-ngs-data}}

Here we will take a closer look at some of the most common NGS data
formats.

    First check you are in the right directory

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n+nb}{pwd}
\end{Verbatim}
\end{tcolorbox}

    It should display something like

\texttt{/home/username/course\_data/data\_formats/data}

Where username will be your username for the server.

    \hypertarget{fasta}{%
\subsection{FASTA}\label{fasta}}

The FASTA format is one of the most basic ways to store sequence data,
and it can be used to store both nucleotide data and protein sequences.
Each sequence in a FASTA file is represented by two parts, a header line
and the actual sequence. The header always starts with the symbol
``\textgreater{}'' and is followed by information about the sequence,
such as a unique identifier. The following lines show two sequences
represented in FASTA format:

\begin{verbatim}
>Sequence_1
CTTGACGACTTGAAAAATGACGAAATCACTAAAAAACGTGAAAAATGAGAAATG
AAAATGACGAAATCACTAAAAAACGTGACGACTTGAAAAATGACCAC
>Sequence_2
CTTGAGACGAAATCACTAAAAAACGTGACGACTTGAAGTGAAAAATGAGAAATG
AAATCATGACGACTTGAAGTGAAAAAGTGAAAAATGAGAAATGAACGTGACGAC
AAAATGACGAAATCATGACGACTTGAAGTGAAAAATAAATGACC
\end{verbatim}

\hypertarget{exercises}{%
\subsubsection{Exercises}\label{exercises}}

\textbf{Q1: How many sequences are there in the fasta file
data/example.fasta? (hint: is there a grep option you can use?)}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    If you get stuck here, do not spend too much time trying to figure this
out and move on. A solution will be provided during the practical
session.

    \hypertarget{fastq}{%
\subsection{FASTQ}\label{fastq}}

FASTQ is a data format for raw unaligned sequencing reads. It is an
extension to the FASTA file format, and includes a quality score for
each base. For paired-end sequencing, two FASTQ files are produced. Have
a look at the example below, containing two reads:

\begin{verbatim}
@ERR007731.739 IL16_2979:6:1:9:1684/1
CTTGACGACTTGAAAAATGACGAAATCACTAAAAAACGTGAAAAATGAGAAATG
+
BBCBCBBBBBBBABBABBBBBBBABBBBBBBBBBBBBBABAAAABBBBB=@>B
@ERR007731.740 IL16_2979:6:1:9:1419/1
AAAAAAAAAGATGTCATCAGCACATCAGAAAAGAAGGCAACTTTAAAACTTTTC
+
BBABBBABABAABABABBABBBAAA>@B@BBAA@4AAA>.>BAA@779:AAA@A
\end{verbatim}

We can see that for each read we get four lines:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  The read metadata, such as the read ID. Starts with \texttt{@} and,
  for paired-end Illumina reads, is terminated with /1 or /2 to show
  that the read is the member of a pair.
\item
  The read
\item
  Starts with \texttt{+} and optionally contains the ID again
\item
  The per base
  \href{https://en.wikipedia.org/wiki/Phred_quality_score}{Phred quality
  score - see https://en.wikipedia.org/wiki/Phred\_quality\_score}
\end{enumerate}

The quality scores range (in theory) from 1 to 94 and are encoded as
\href{https://en.wikipedia.org/wiki/ASCII}{ASCII characters - see
https://en.wikipedia.org/wiki/ASCII)}. The first 32 ASCII codes are
reserved for control characters which are not printable, and the 33rd is
reserved for space. Neither of these can be used in the quality string,
so we need to subtract 33 from whatever the value of the quality
character is. For example, the ASCII code of ``A'' is 65, so the
corresponding quality is:

\begin{verbatim}
Q = 65 - 33 = 32
\end{verbatim}

The Phred quality score \texttt{Q} relates to the base-calling error
probability \texttt{P} as

~~~~~~ P = 10-Q/10

The Phred quality score is a measure of the quality of base calls. For
example, a base assigned with a Phred quality score of 30 tells us that
there is a 1 in 1000 chance that this base was called incorrectly.

\begin{longtable}[]{@{}lll@{}}
\toprule
Phred Quality Score & Probability of incorrect base call & Base call
accuracy \\
\midrule
\endhead
10 & 1 in 10 & 90\% \\
20 & 1 in 100 & 99\% \\
30 & 1 in 1000 & 99.9\% \\
40 & 1 in 10,000 & 99.99\% \\
50 & 1 in 100,000 & 99.999\% \\
60 & 1 in 1,000,000 & 99.9999\% \\
\bottomrule
\end{longtable}

The following simple perl command will print the quality score value for
an ASCII character. Try changing the ``A'' to another character, for
example one from the quality strings above (e.g.~@, = or B).

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
perl\PY{+w}{ }\PYZhy{}e\PY{+w}{ }\PY{l+s+s1}{\PYZsq{}printf \PYZdq{}\PYZpc{}d\PYZbs{}n\PYZdq{},ord(\PYZdq{}A\PYZdq{})\PYZhy{}33;\PYZsq{}}
\end{Verbatim}
\end{tcolorbox}

    Something to be aware of is that two different ways of calculating the
quality scores have historically been in use. The standard Sanger
variant uses the Phred score, while the Solexa pipeline earlier used a
\href{https://en.wikipedia.org/wiki/FASTQ_format\#Quality}{different
mapping}.

\hypertarget{exercises}{%
\subsubsection{Exercises}\label{exercises}}

\textbf{Q2: How many reads are there in the file example.fastq? (Hint:
remember that \texttt{@} is a possible quality score. Is there something
else in the header that is unique?)}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    Again, don't worry if you cannot solve this the solution will be
provided during the practical session.

    \hypertarget{sambam}{%
\subsection{SAM/BAM}\label{sambam}}

\href{https://samtools.github.io/hts-specs/SAMv1.pdf}{SAM (Sequence
Alignment/Map)} format was developed by the 1000 Genomes Project group
in 2009 and is a unified format for storing read alignments to a
reference genome. BAM is the compressed version of SAM. SAM/BAM format
is the accepted standard format for storing NGS sequencing reads, base
qualities, associated meta-data and alignments of the data to a
reference genome. If no reference genome is available, the data can also
be stored unaligned.

The files consist of a header section (optional) and an alignment
section. The alignment section contains one record (a single DNA
fragment alignment) per line describing the alignment between fragment
and reference. Each record has 11 fixed columns and optional
key:type:value tuples. Open the
\href{https://samtools.github.io/hts-specs/SAMv1.pdf}{SAM/BAM file
specification document} as you may need to refer to it throughout this
tutorial.

Now let us have a closer look at the different parts of the SAM/BAM
format.

\hypertarget{header-section}{%
\subsubsection{Header Section}\label{header-section}}

Each line in the SAM header begins with an \texttt{@}, followed by a
two-letter header record type code as defined in the
\href{https://samtools.github.io/hts-specs/SAMv1.pdf}{SAM/BAM format
specification document}. Each record type can contain meta-data captured
as a series of key-value pairs in the format of `TAG:VALUE'.

\hypertarget{read-groups}{%
\paragraph{Read groups}\label{read-groups}}

One useful record type is RG which can be used to describe each unit of
sequencing, this can be a lane of sequencing data. The RG code can be
used to capture extra meta-data for the sequencing lane. Some common RG
TAGs are:

\begin{itemize}
\tightlist
\item
  ID: SRR/ERR number
\item
  PL: Sequencing platform
\item
  PU: Run name
\item
  LB: Library name
\item
  PI: Insert fragment size
\item
  SM: Individual/Sample
\item
  CN: Sequencing centre
\end{itemize}

While most of these are self explanitory, insert fragment size may
occasionally be negative. This simply indicates that the reads found are
overlapping while its size is less than 2 x read length.

    \hypertarget{exercises}{%
\subsubsection{Exercises}\label{exercises}}

From reading section 1.3 of the SAM specification, look at the following
line from the header of the SAM/BAM file:

\begin{verbatim}
@RG ID:ERR003612 PL:ILLUMINA LB:g1k-sc-NA20538-TOS-1 PI:2000 DS:SRP000540 SM:NA20538 CN:SC
\end{verbatim}

\textbf{Q3: What does RG stand for?}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \textbf{Q4: What is the sequencing platform?}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \textbf{Q5: What is the sequencing centre?}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \textbf{Q6: What is the expected fragment insert size?}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \hypertarget{alignment-section}{%
\subsubsection{Alignment Section}\label{alignment-section}}

The alignment section of SAM files contains one line per fragment
alignment, which in turn contains the columns listed below. The first 11
columns are mandatory.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  QNAME: Query NAME of the read or the read pair
\item
  FLAG: Bitwise FLAG (pairing, strand, mate strand, etc.)
\item
  RNAME: Reference sequence NAME
\item
  POS: 1-Based leftmost POSition of clipped alignment
\item
  MAPQ: MAPping Quality (Phred-scaled)
\item
  CIGAR: Extended CIGAR string (operations: MIDNSHPX=)
\item
  MRNM: Mate Reference NaMe ('=' if same as RNAME)
\item
  MPOS: 1-Based leftmost Mate POSition
\item
  ISIZE: Inferred Insert SIZE
\item
  SEQ: Query SEQuence on the same strand as the reference
\item
  QUAL: Query QUALity (ASCII-33=Phred base quality)
\item
  OTHER: Optional fields
\end{enumerate}

The image below provides a visual guide to some of the columns of the
SAM format.

    \begin{figure}
\centering
\includegraphics{img/SAM_BAM.png}
\caption{SAM format}
\end{figure}

    In a SAM file, this image representation could for instance translate to
the following entry of 100 bases:

\texttt{ERR005816.1408831\ \ 163\ Chr1\ \ \ \ 19999970\ \ \ \ 23\ \ 40M5D30M2I28M\ \ \ =\ \ \ 20000147\ \ \ \ 213\ GGTGGGTGGATCACCTGAGATCGGGAGTTTGAGACTAGGTGG...\ \ \ \ \textless{}=@A@??@=@A@A\textgreater{}@BAA@ABA:\textgreater{}@\textless{}\textgreater{}=BBB9@@2B3\textless{}=@A@...}

\hypertarget{exercises}{%
\subsubsection{Exercises}\label{exercises}}

Let's have a look at example.sam. This file contains only a subset of
the alignment section of a BAM-file that we will look closer at soon.
Notice that we can use the standard UNIX operations like \textbf{cat} on
this file.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
cat\PY{+w}{ }example.sam
\end{Verbatim}
\end{tcolorbox}

    \textbf{Q8: What is the mapping quality of ERR003762.5016205? (Hint: can
you use grep and awk to find this?)}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \textbf{Q9: What is the CIGAR string for ERR003814.6979522? (We will go
through the meaning of CIGAR strings in the next section)}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \textbf{Q10: What is the inferred insert size?}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \hypertarget{cigar-string}{%
\subsubsection{CIGAR string}\label{cigar-string}}

Column 6 of the alignment is the CIGAR string for that alignment. The
CIGAR string provides a compact representation of sequence alignment.
Have a look at the table below. It contains the meaning of all different
symbols of a CIGAR string:

\begin{longtable}[]{@{}ll@{}}
\toprule
Symbol & Meaning \\
\midrule
\endhead
M & alignment match or mismatch \\
= & sequence match \\
X & sequence mismatch \\
I & insertion to the reference \\
D & deletion from the reference \\
S & soft clipping (clipped sequences present in SEQ) \\
H & hard clipping (clipped sequences NOT present in SEQ) \\
N & skipped region from the reference \\
P & padding (silent deletion from padded reference) \\
\bottomrule
\end{longtable}

Below are two examples describing the CIGAR string in more detail.

\textbf{Example 1:}\\
Ref:~~~~~ACGTACGTACGTACGT\\
Read:~~ACGT-~-~-~-~ACGTACGA\\
Cigar: 4M 4D 8M

The first four bases in the read are the same as in the reference, so we
can represent these as 4M in the CIGAR string. Next comes 4 deletions,
represented by 4D, followed by 7 alignment matches and one alignment
mismatch, represented by 8M. Note that the mismatch at position 16 is
included in 8M. This is because it still aligns to the reference.

\textbf{Example 2:}\\
Ref:~~~~~ACTCAGTG-~-~GT\\
Read:~~ACGCA-~TGCAGTtagacgt\\
Cigar: 5M 1D 2M 2I 2M 7S

Here we start off with 5 alignment matches and mismatches, followed by
one deletion. Then we have two more alignment matches, two insertions
and two more matches. At the end, we have seven soft clippings, 7S.
These are clipped sequences that are present in the SEQ (Query SEQuence
on the same strand as the reference).

\hypertarget{exercises}{%
\subsubsection{Exercises}\label{exercises}}

\textbf{Q11: What does the CIGAR from Q9 mean?}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \textbf{Q12: How would you represent the following alignment with a
CIGAR string?}

Ref:~~~~~ACGT-~-~-~-~ACGTACGT\\
Read:~~ACGTACGTACGTACGT

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \hypertarget{flags}{%
\subsubsection{Flags}\label{flags}}

Column 2 of the alignment contains a combination of bitwise FLAGs
describing the alignment. The following table contains the information
you can get from the bitwise FLAGs:

\begin{longtable}[]{@{}llll@{}}
\toprule
Hex & Dec & Flag & Description \\
\midrule
\endhead
0x1 & 1 & PAIRED & paired-end (or multiple-segment) sequencing
technology \\
0x2 & 2 & PROPER\_PAIR & each segment properly aligned according to the
aligner \\
0x4 & 4 & UNMAP & segment unmapped \\
0x8 & 8 & MUNMAP & next segment in the template unmapped \\
0x10 & 16 & REVERSE & SEQ is reverse complemented \\
0x20 & 32 & MREVERSE & SEQ of the next segment in the template is
reversed \\
0x40 & 64 & READ1 & the first segment in the template \\
0x80 & 128 & READ2 & the last segment in the template \\
0x100 & 256 & SECONDARY & secondary alignment \\
0x200 & 512 & QCFAIL & not passing quality controls \\
0x400 & 1024 & DUP & PCR or optical duplicate \\
0x800 & 2048 & SUPPLEMENTARY & supplementary alignment \\
\bottomrule
\end{longtable}

For example, if you have an alignment with FLAG set to 113, this can
only be represented by decimal codes \texttt{64\ +\ 32\ +\ 16\ +\ 1}, so
we know that these four flags apply to the alignment and the alignment
is paired-end, reverse complemented, sequence of the next template/mate
of the read is reversed and the read aligned is the first segment in the
template.

\hypertarget{primary-secondary-and-supplementary-alignments}{%
\paragraph{Primary, secondary and supplementary
alignments}\label{primary-secondary-and-supplementary-alignments}}

A read that aligns to a single reference sequence (including insertions,
deletions, skips and clipping but not direction changes), is a
\textbf{linear alignment}. If a read cannot be represented as a linear
alignment, but instead is represented as a group of linear alignments
without large overlaps, it is called a \textbf{chimeric alignment}.
These can for instance be caused by structural variations. Usually, one
of the linear alignments in a chimeric alignment is considered to be the
\textbf{representative} alignment, and the others are called
\textbf{supplementary}.

Sometimes a read maps equally well to more than one spot. In these
cases, one of the possible alignments is marked as the \textbf{primary}
alignment and the rest are marked as \textbf{secondary} alignments.

\hypertarget{bam}{%
\subsubsection{BAM}\label{bam}}

BAM (Binary Alignment/Map) format, is a binary version of SAM. This
means that, while SAM is human readable, BAM is only readable for
computers. BAM was developed for fast processing and random access. To
achieve this, BGZF (Block GZIP) compression is used for indexing. BAM
files can be viewed using samtools, and will then have the same format
as a SAM file. The key features of BAM are:

\begin{itemize}
\tightlist
\item
  Can store alignments from most mappers
\item
  Supports multiple sequencing technologies
\item
  Supports indexing for quick retrieval/viewing
\item
  Compact size (e.g.~112Gbp Illumina = 116GB disk space)
\item
  Reads can be grouped into logical groups e.g.~lanes, libraries,
  samples
\item
  Widely supported by variant calling packages and viewers
\end{itemize}

    \hypertarget{exercises}{%
\subsubsection{Exercises}\label{exercises}}

Since BAM is a binary format, we can't use the standard Linux operations
(cat, less, head, grep etc.) directly on this file format.
\textbf{Samtools} is a set of programs for interacting with SAM and BAM
files. Using the samtools view command, print the header of the BAM
file:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
samtools\PY{+w}{ }view\PY{+w}{ }\PYZhy{}H\PY{+w}{ }NA20538.bam
\end{Verbatim}
\end{tcolorbox}

    \textbf{Q13: What version of the human assembly was used to perform the
alignments? (Hint: Can you spot this somewhere in the @SQ records?)}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \textbf{Q14: How many sequencing runs/lanes are in this BAM file? (Hint:
Do you recall what RG represents?)}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \textbf{Q15: What programs were used to create this BAM file? (Hint:
have a look for the program record, @PG)}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \textbf{Q16: What version of bwa was used to align the reads? (Hint: is
there anything in the @PG record that looks like it could be a version
tag?)}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    The output from running samtools view on a BAM file without any options
is a headerless SAM file. This gets printed to STDOUT in the terminal,
so we will want to pipe it to something. Let's have a look at the first
read of the BAM file:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
samtools\PY{+w}{ }view\PY{+w}{ }NA20538.bam\PY{+w}{ }\PY{p}{|}\PY{+w}{ }head\PY{+w}{ }\PYZhy{}n\PY{+w}{ }\PY{l+m}{1}
\end{Verbatim}
\end{tcolorbox}

    \textbf{Q17: What is the name of the first read? (Hint: have a look at
the \href{formats.ipynb\#Alignment-Section}{alignment section} if you
can't recall the different fields)}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \textbf{Q18: What position does the alignment of the read start at?}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \hypertarget{cram}{%
\subsection{CRAM}\label{cram}}

Even though BAM files are compressed, they are still too large.
Typically they use 1.5-2 bytes for each base pair of sequencing data
that they contain, and while disk capacity is ever improving, increases
in disk capacity are being far outstripped by sequencing technologies.

    \begin{figure}
\centering
\includegraphics{img/compression_cram.png}
\caption{Growth of DNA sequencing}
\end{figure}

    BAM stores all of the data, this includes every read base, every base
quality, and it uses a single conventional compression technique for all
types of data. Therefore, CRAM was designed for better compression of
genomic data than SAM/BAM. CRAM uses three important concepts:

\begin{itemize}
\tightlist
\item
  Reference based compression
\item
  Controlled loss of quality information
\item
  Different compression methods to suit the type of data, e.g.~base
  qualities vs.~metadata vs.~extra tags
\end{itemize}

The figure below displays how reference-based compression works. Instead
of saving all the bases of all the reads, only the nucleotides that
differ from the reference, and their positions, are kept.

    \begin{figure}
\centering
\includegraphics{img/CRAM_format.png}
\caption{CRAM1}
\end{figure}

    \begin{figure}
\centering
\includegraphics{img/CRAM_format2.png}
\caption{CRAM2}
\end{figure}

    In lossless (no information is lost) mode a CRAM file is 60\% of the
size of a BAM file, so archives and sequencing centres are now moving
from BAM to CRAM.

Since samtools 1.3, CRAM files can be read in the same way that BAM
files can. We will look closer at how you can convert between SAM, BAM
and CRAM formats in the next section.

    \hypertarget{indexing}{%
\subsection{Indexing}\label{indexing}}

To allow for fast random access of regions in BAM and CRAM files, they
can be indexed. The files must first be coordinate-sorted. This can be
done using \textbf{samtools sort}. If no options are supplied, it will
by default sort by the left-most position.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
samtools\PY{+w}{ }sort\PY{+w}{ }\PYZhy{}o\PY{+w}{ }NA20538\PYZus{}sorted.bam\PY{+w}{ }NA20538.bam
\end{Verbatim}
\end{tcolorbox}

    Now we can use \textbf{samtools index} to create an index file (.bai)
for our sorted BAM file:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
samtools\PY{+w}{ }index\PY{+w}{ }NA20538\PYZus{}sorted.bam
\end{Verbatim}
\end{tcolorbox}

    To look for reads mapped to a specific region, we can use
\textbf{samtools view} and specify the region we are interested in as:
RNAME{[}:STARTPOS{[}-ENDPOS{]}{]}. For example, if we wanted to look at
all the reads mapped to a region called chr4, we could use:

\begin{verbatim}
samtools view alignment.bam chr4
\end{verbatim}

To look at the region on chr4 beginning at position 1,000,000 and ending
at the end of the chromosome, we can do:

\begin{verbatim}
samtools view alignment.bam chr4:1000000
\end{verbatim}

And to explore the 1001bp long region on chr4 beginning at position
1,000 and ending at position 2,000, we can use:

\begin{verbatim}
samtools view alignment.bam chr4:1000-2000
\end{verbatim}

    \hypertarget{exercises}{%
\subsubsection{Exercises}\label{exercises}}

\textbf{Q19: How many reads are mapped to region 20025000-20030000 on
chromosome 1?}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \hypertarget{vcfbcf}{%
\subsection{VCF/BCF}\label{vcfbcf}}

The VCF file format and its binary version BCF were introduced to store
variation data. VCF consists of tab-delimited text and is parsable by
standard UNIX commands which makes it flexible and user-extensible. The
figure below provides an overview of the different components of a VCF
file:

    \begin{figure}
\centering
\includegraphics{img/VCF1.png}
\caption{VCF format}
\end{figure}

    \hypertarget{vcf-header}{%
\subsubsection{VCF header}\label{vcf-header}}

The VCF header consists of meta-information lines (starting with
\texttt{\#\#}) and a header line (starting with \texttt{\#}). All
meta-information lines are optional and can be put in any order, except
for \textit{fileformat}. This holds the information about which version of
VCF is used and must come first.

The meta-information lines consist of key=value pairs. Examples of
meta-information lines that can be included are \#\#INFO, \#\#FORMAT and
\#\#reference. The values can consist of multiple fields enclosed by
\texttt{\textless{}\textgreater{}}. More information about these fields
is available in the
\href{http://samtools.github.io/hts-specs/VCFv4.3.pdf}{VCF
specification}.

\hypertarget{header-line}{%
\subsubsection{Header line}\label{header-line}}

The header line starts with \texttt{\#} and consists of 8 required
fields:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  CHROM: an identifier from the reference genome
\item
  POS: the reference position
\item
  ID: a list of unique identifiers (where available)
\item
  REF: the reference base(s)
\item
  ALT: the alternate base(s)
\item
  QUAL: a phred-scaled quality score
\item
  FILTER: filter status
\item
  INFO: additional information
\end{enumerate}

If the file contains genotype data, the required fields are also
followed by a FORMAT column header, and then a number of sample IDs. The
FORMAT field specifies the data types and order. Some examples of these
data types are:

\begin{itemize}
\tightlist
\item
  GT: Genotype, encoded as allele values separated by either of / or
  \textbar{}
\item
  DP: Read depth at this position for this sample
\item
  GQ: Conditional genotype quality, encoded as a phred quality
\end{itemize}

\hypertarget{body}{%
\subsubsection{Body}\label{body}}

In the body of the VCF, each row contains information about a position
in the genome along with genotype information on samples for each
position, all according to the fields in the header line.

\hypertarget{bcf}{%
\subsubsection{BCF}\label{bcf}}

VCF can be compressed with BGZF (bgzip) and indexed with TBI or CSI
(tabix), but even compressed it can still be very big. For example, a
compressed VCF with 3781 samples of human data will be 54 GB for
chromosome 1, and 680 GB for the whole genome.

VCFs can also be slow to parse, as text conversion is slow. The main
bottleneck is the ``FORMAT'' fields. For this reason the BCF format, a
binary representation of VCF, was developed. In BCF files the fields are
rearranged for fast access. The following images show the process of
converting a VCF file into a BCF file.

    \begin{figure}
\centering
\includegraphics{img/VCF2.png}
\caption{VCF2}
\end{figure}

    \begin{figure}
\centering
\includegraphics{img/VCF3.png}
\caption{VCF3}
\end{figure}

    Bcftools comprises a set of programs for interacting with VCF and BCF
files. It can be used to convert between VCF and BCF and to view or
extract records from a region.

\hypertarget{bcftools-view}{%
\paragraph{bcftools view}\label{bcftools-view}}

Let's have a look at the header of the file 1kg.bcf in the data
directory. Note that bcftools uses \textbf{\texttt{-h}} to print only
the header, while samtools uses \textbf{\texttt{-H}} for this.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bcftools\PY{+w}{ }view\PY{+w}{ }\PYZhy{}h\PY{+w}{ }1kg.bcf
\end{Verbatim}
\end{tcolorbox}

    Similarly to BAM, BCF supports random access, that is, fast retrieval
from a given region. For this, the file must be indexed:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bcftools\PY{+w}{ }index\PY{+w}{ }1kg.bcf
\end{Verbatim}
\end{tcolorbox}

    Now we can extract all records from the region 20:24042765-24043073,
using the \textbf{\texttt{-r}} option. The \textbf{\texttt{-H}} option
will make sure we don't include the header in the output:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bcftools\PY{+w}{ }view\PY{+w}{ }\PYZhy{}H\PY{+w}{ }\PYZhy{}r\PY{+w}{ }\PY{l+m}{20}:24042765\PYZhy{}24043073\PY{+w}{ }1kg.bcf
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{bcftools-query}{%
\paragraph{bcftools query}\label{bcftools-query}}

The versatile \textbf{bcftools query} command can be used to extract any
VCF field. Combined with standard UNIX commands, this gives a powerful
tool for quick querying of VCFs. Have a look at the usage options:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bcftools\PY{+w}{ }query\PY{+w}{ }\PYZhy{}h
\end{Verbatim}
\end{tcolorbox}

    Let's try out some useful options. As you can see from the usage,
\textbf{\texttt{-l}} will print a list of all the samples in the file.
Give this a go:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bcftools\PY{+w}{ }query\PY{+w}{ }\PYZhy{}l\PY{+w}{ }1kg.bcf
\end{Verbatim}
\end{tcolorbox}

    Another very useful option is \textbf{\texttt{-s}} which allows you to
extract all the data relating to a particular sample. This is a
\href{http://samtools.github.io/bcftools/bcftools.html\#common_options}{common
option} meaning it can be used for many bcftools commands, like
\texttt{bcftools\ view}. Try this for sample HG00131:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bcftools\PY{+w}{ }view\PY{+w}{ }\PYZhy{}s\PY{+w}{ }HG00131\PY{+w}{ }1kg.bcf\PY{+w}{ }\PY{p}{|}\PY{+w}{ }head\PY{+w}{ }\PYZhy{}n\PY{+w}{ }\PY{l+m}{50}
\end{Verbatim}
\end{tcolorbox}

    The format option, \textbf{\texttt{-f}} can be used to select what gets
printed from your query command. For example, the following will print
the position, reference base and alternate base for sample HG00131,
separated by tabs:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bcftools\PY{+w}{ }query\PY{+w}{ }\PYZhy{}f\PY{l+s+s1}{\PYZsq{}\PYZpc{}POS\PYZbs{}t\PYZpc{}REF\PYZbs{}t\PYZpc{}ALT\PYZbs{}n\PYZsq{}}\PY{+w}{ }\PYZhy{}s\PY{+w}{ }HG00131\PY{+w}{ }1kg.bcf\PY{+w}{ }\PY{p}{|}\PY{+w}{ }head
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{exercises}{%
\subsubsection{Exercises}\label{exercises}}

Now, try and answer the following questions about the file 1kg.bcf in
the data directory. For more information about the different usage
options you can open the
\href{http://samtools.github.io/bcftools/bcftools.html\#query}{bcftools
query manual page -
http://samtools.github.io/bcftools/bcftools.html\#query)} in a new tab.

    \textbf{Q20: What version of the human assembly do the coordinates refer
to?}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \textbf{Q21: How many samples are there in the BCF?}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \textbf{Q22: What is the genotype of the sample HG00107 at the position
20:24019472? (Hint: use the combination of -r, -s, and -f options)}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    Now continue to the next section of the tutorial:
\href{assessment.ipynb}{QC assement}.


    % Add a bibliography block to the postdoc



\newpage





    \hypertarget{qc-assessment-of-ngs-data}{%
\section{QC assessment of NGS data}\label{qc-assessment-of-ngs-data}}

As mentioned previously, QC is an important part of any analysis. In
this section we are going to look at some of the metrics and graphs that
can be used to assess the QC of NGS data.

\hypertarget{base-quality}{%
\subsection{Base quality}\label{base-quality}}

\href{https://en.wikipedia.org/wiki/Illumina_dye_sequencing}{Illumina
sequencing} technology relies on sequencing by synthesis. One of the
most common problems with this is \textbf{dephasing}. For each
sequencing cycle, there is a possibility that the replication machinery
slips and either incorporates more than one nucleotide or perhaps misses
to incorporate one at all. The more cycles that are run (i.e.~the longer
the read length gets), the greater the accumulation of these types of
errors gets. This leads to a heterogeneous population in the cluster,
and a decreased signal purity, which in turn reduces the precision of
the base calling. The figure below shows an example of this.

    \begin{figure}
\centering
\includegraphics{img/base_qual.png}
\caption{Base Quality}
\end{figure}

    Because of dephasing, it is possible to have high-quality data at the
beginning of the read but really low-quality data towards the end of the
read. In those cases you can decide to trim off the low-quality reads,
for example using a tool called
\href{http://www.usadellab.org/cms/?page=trimmomatic}{Trimmomatic}.

The figures below shows an example of a high-quality read data (top) and
a poor quality read data (bottom).

    \begin{figure}
\centering
\includegraphics{img/base_qual_pass.png}
\caption{High-quality read data}
\end{figure}

    \begin{figure}
\centering
\includegraphics{img/base_qual_fail.png}
\caption{Poor quality read data}
\end{figure}

    \hypertarget{other-base-calling-errors}{%
\subsection{Other base calling errors}\label{other-base-calling-errors}}

There are several different reasons for a base to be called incorrectly,
as shown in the figure below. \textbf{Phasing noise} and \textbf{signal
decay} is a result of the dephasing issue described above. During
library preparation, \textbf{mixed clusters} can occur if multiple
templates get co-located. These clusters should be removed from the
downstream analysis. \textbf{Boundary effects} occur due to optical
effects when the intensity is uneven across each tile, resulting in
higher intensity found toward the center. \textbf{Cross-talk} occurs
because the emission frequency spectra for each of the four base dyes
partly overlap, creating uncertainty. Finally, for previous sequencing
cycle methods \textbf{T fluorophore accumulation} was an issue, where
incomplete removal of the dye coupled to thymine lead to an ambient
accumulation the nucleotides, causing a false high Thymine trend.

    \begin{figure}
\centering
\includegraphics{img/base_calling_errors.jpg}
\caption{Base Calling Errors}
\end{figure}

    \textit{Base-calling for next-generation sequencing platforms}, doi:
\href{https://academic.oup.com/bib/article/12/5/489/268399}{10.1093/bib/bbq077}

\hypertarget{mismatches-per-cycle}{%
\subsection{Mismatches per cycle}\label{mismatches-per-cycle}}

Aligning reads to a high-quality reference genome can provide insight to
the quality of a sequencing run by showing you the mismatches to the
reference sequence. This can help you detect cycle-specific errors.
Mismatches can occur due to two main causes, sequencing errors and
differences between your sample and the reference genome, which is
important to bear in mind when interpreting mismatch graphs. The figures
below show an example of a good run (top) and a bad one (bottom). In the
first figure, the distribution of the number of mismatches is even
between the cycles, which is what we would expect from a good run.
However, in the second figure, two cycles stand out with a lot of
mismatches compared to the other cycles.

    \begin{figure}
\centering
\includegraphics{img/mismatch_per_cycle_pass.png}
\caption{Good run}
\end{figure}

    \begin{figure}
\centering
\includegraphics{img/mismatch_per_cycle_fail.png}
\caption{Poor run}
\end{figure}

    \hypertarget{gc-content}{%
\subsection{GC content}\label{gc-content}}

It is a good idea to compare the GC content of the reads against the
expected distribution in a reference sequence. The GC content varies
between species, so a shift in GC content like the one seen below could
be an indication of sample contamination. In the left image below, we
can see that the GC content of the sample is about the same as for the
reference, at \textasciitilde38\%. However, in the right figure, the GC
content of the sample is closer to 55\%, indicating that there is an
issue with this sample.

    \begin{figure}
\centering
\includegraphics{img/gc_bias.png}
\caption{GC Bias}
\end{figure}

    \hypertarget{gc-content-by-cycle}{%
\subsection{GC content by cycle}\label{gc-content-by-cycle}}

Looking at the GC content per cycle can help detect if the adapter
sequence was trimmed. For a random library, it is expected to be little
to no difference between the different bases of a sequence run, so the
lines in this plot should be parallel with each other like in the first
of the two figures below. In the second of the figures, the initial
spikes are likely due to adapter sequences that have not been removed.

    \begin{figure}
\centering
\includegraphics{img/acgt_per_cycle_pass.png}
\caption{Good run}
\end{figure}

    \begin{figure}
\centering
\includegraphics{img/acgt_per_cycle_fail.png}
\caption{Poor run}
\end{figure}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{c+c1}{\PYZsh{}\PYZsh{} GC Bias/Depth}
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{insert-size}{%
\subsection{Insert size}\label{insert-size}}

For paired-end sequencing the size of DNA fragments also matters. In the
first of the examples below, the insert size peaks around 440 bp. In the
second however, there is also a peak at around 200 bp. This indicates
that there was an issue with the fragment size selection during library
prep.

    \begin{figure}
\centering
\includegraphics{img/insert_size_pass.png}
\caption{Good run}
\end{figure}

    \begin{figure}
\centering
\includegraphics{img/insert_size_fail.png}
\caption{Poor run}
\end{figure}

    \hypertarget{exercises}{%
\subsubsection{Exercises}\label{exercises}}

\textbf{Q1: The figure below is from a 100bp paired-end sequencing. Can
you spot any problems?}

    \begin{figure}
\centering
\includegraphics{img/insert_size_quiz.png}
\caption{Q1}
\end{figure}

    \hypertarget{insertionsdeletions-per-cycle}{%
\subsection{Insertions/Deletions per
cycle}\label{insertionsdeletions-per-cycle}}

Sometimes, air bubbles occur in the flow cell, which can manifest as
false indels. The spike in the second image provides an example of how
this can look.

    \begin{figure}
\centering
\includegraphics{img/indels-per-cycle.pass.png}
\caption{Good run}
\end{figure}

    \begin{figure}
\centering
\includegraphics{img/indels-per-cycle.fail.png}
\caption{Poor run}
\end{figure}

    \hypertarget{generating-qc-stats}{%
\subsection{Generating QC stats}\label{generating-qc-stats}}

Now let's try this out! We will generate QC stats for two lanes of
Illumina paired-end sequencing data from yeast. We will use the bwa
mapper to align the data to the
\href{ftp://ftp.ensembl.org/pub/current_fasta/saccharomyces_cerevisiae/dna}{Saccromyces
cerevisiae genome}, followed by samtools stats to generate the stats.

Read pairs are usually stored in two separate FASTQ files so that n-th
read in the first file and the n-th read in the second file constitute a
read pair. Can you devise a quick sanity check that reads in these two
files indeed form pairs? The files must have the same number of lines
and the naming of the reads usually suggests if they form a pair. The
location of the files is:

\begin{verbatim}
lane1/s_7_1.fastq
lane1/s_7_2.fastq
\end{verbatim}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    Let's have a look at the script we are going to run to create the
mappings:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
cat\PY{+w}{ }create\PYZus{}mapping.sh
\end{Verbatim}
\end{tcolorbox}

    TDon't worry about he details of this for now as we will cover mapping
data to a reference genome later in the course).

Now run the script to create the mappings and stats:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
./create\PYZus{}mapping.sh
\end{Verbatim}
\end{tcolorbox}

    The script will produce the BAM file lane1.sorted.bam and a matching
index file:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ls\PY{+w}{ }\PYZhy{}alrt
\end{Verbatim}
\end{tcolorbox}

    Now we will use \textbf{\texttt{samtools\ stats}} to generate the stats
for the primary alignments. The option \textbf{\texttt{-f}} can be used
to filter reads with specific tags, while \textbf{\texttt{-F}} can be
used to \textit{filter out} reads with specific tags. The following
command will include only primary alignments:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
samtools\PY{+w}{ }stats\PY{+w}{ }\PYZhy{}F\PY{+w}{ }SECONDARY\PY{+w}{ }lane1.sorted.bam\PY{+w}{ }\PY{l+s+se}{\PYZbs{}}
\PY{+w}{    }\PYZgt{}\PY{+w}{ }lane1.sorted.bam.bchk
\end{Verbatim}
\end{tcolorbox}

    Have a look at the first 47 lines of the statistics file that was
generated:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
head\PY{+w}{ }\PYZhy{}n\PY{+w}{ }\PY{l+m}{47}\PY{+w}{ }lane1.sorted.bam.bchk
\end{Verbatim}
\end{tcolorbox}

    This file contains a number of useful stats that we can use to get a
better picture of our data, and it can even be plotted with
\textbf{\texttt{plot-bamstats}}, as you will see soon. First let's have
a closer look at some of the different stats. Each part of the file
starts with a \texttt{\#} followed by a description of the section and
how to extract it from the file. Let's have a look at all the sections
in the file:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
grep\PY{+w}{ }\PYZca{}\PY{l+s+s1}{\PYZsq{}\PYZsh{}\PYZsq{}}\PY{+w}{ }lane1.sorted.bam.bchk\PY{+w}{ }\PY{p}{|}\PY{+w}{ }grep\PY{+w}{ }\PY{l+s+s1}{\PYZsq{}Use\PYZsq{}}
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{summary-numbers-sn}{%
\subsubsection{Summary Numbers (SN)}\label{summary-numbers-sn}}

This initial section contains a summary of the alignment and includes
some general statistics. In particular, you can see how many bases
mapped, and how much of the genome that was covered.

    Now look at the output and try to answer the questions below.

\textbf{Q2: What is the total number of reads?}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \textbf{Q3: What proportion of the reads were mapped?}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \textbf{Q4: How many pairs were mapped to a different chromosome?}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \textbf{Q5: What is the insert size mean and standard deviation?}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \textbf{Q6: How many reads were paired properly?}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    Finally, we will create some QC plots from the output of the stats
command using the command \textbf{plot-bamstats} which is included in
the samtools package:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
plot\PYZhy{}bamstats\PY{+w}{ }\PYZhy{}p\PY{+w}{ }lane1\PYZhy{}plots/\PY{+w}{ }lane1.sorted.bam.bchk
\end{Verbatim}
\end{tcolorbox}

    Now in your web browser open the file lane1-plots/index.html to view the
QC information.

\textbf{Q7: How many reads have zero mapping quality?}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \textbf{Q8: Which read (forward/reverse) of the first fragments and
second fragments are higher base quality on average?}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \hypertarget{fastqc}{%
\subsection{FastQC}\label{fastqc}}

An alternative method to asseing the QC of your sequnce data is to the
the FastQC software application. This is a usefull tool to get an
initial assement of the quality of your sequence data without having to
align/map your data to a reference genome. It provide similar statistics
and graphs as above and additional information like estimates of
duplicated sequnces, adapter contamination, number of N's in the
sequence runs. Let's take a look.

To run FastQC on XXX

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
fastqc\PY{+w}{ }*.fastq.gz\PY{+w}{ }\PYZhy{}o\PY{+w}{ }fastqc\PYZus{}results
\end{Verbatim}
\end{tcolorbox}

    Let's look at the output files produced:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ls\PY{+w}{ }fastqc\PYZus{}results
\end{Verbatim}
\end{tcolorbox}

    As you can see FastQC runs on each fastq file seperately and generates
QC stats and an html web page for each fastq file.

    Open the html file for xxx.fastq.gz

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
firefox\PY{+w}{ }XXX\PY{+w}{ }\PY{p}{\PYZam{}}
\end{Verbatim}
\end{tcolorbox}

    Ask questions about the results:

    Often you will hace sequence data for multiple samples or multiple runs
and it will be useful to look at the quality reports for all data in one
combined report. This combined report can be generated with a software
application called multiqc. Let;s gather together all the fastqc report
files from above into a single report.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
multiqc\PY{+w}{ }fastqc\PYZus{}results
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ls
\end{Verbatim}
\end{tcolorbox}

    This generates a singel report called multiqc\_report.html. Let's look
at this report:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
firefox\PY{+w}{ }multiqc\PYZus{}report.html\PY{+w}{ }\PY{p}{\PYZam{}}
\end{Verbatim}
\end{tcolorbox}

    What is your assesmment of this sequencing run? Is it good quality data?

    Now continue to the next section of the tutorial:
\href{contamination.ipynb}{Identifying contamination}


    % Add a bibliography block to the postdoc



\newpage





    \hypertarget{identifying-contamination}{%
\section{Identifying contamination}\label{identifying-contamination}}

It is always a good idea to check that your data is from the species you
expect it to be.

    \hypertarget{bactinspector}{%
\subsection{Bactinspector}\label{bactinspector}}

One useful software application for doing this is bactinspector.
Bactinspector\ldots.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bactinspector\PY{+w}{ }check\PYZus{}species\PY{+w}{ }\PYZhy{}fq\PY{+w}{ }13681\PYZus{}1\PYZsh{}18\PYZus{}1.fastq.gz
\end{Verbatim}
\end{tcolorbox}

    What is the predicted species?

    Another useful tool for identifying contamination is
\href{https://www.ebi.ac.uk/research/enright/software/kraken}{Kraken}.

\hypertarget{installing-kraken}{%
\subsection{Installing kraken}\label{installing-kraken}}

Up until now all the software you required has been available on the
virtual machine as it has been installed via conda (a software package
manager). Kraken is not available on the virtual machine so for this
part of the practical we will attempt to use conda to install kraken:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
conda\PY{+w}{ }create\PY{+w}{ }\PYZhy{}n\PY{+w}{ }kraken\PY{+w}{ }\PY{n+nv}{kraken}\PY{o}{=}x.y
\end{Verbatim}
\end{tcolorbox}

    Once installed sucessfully activate the software environment that
contains kraken.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
conda\PY{+w}{ }activate\PY{+w}{ }kraken
\end{Verbatim}
\end{tcolorbox}

    The reason we have to activate the environment is that often
bioinformatics software has multiple software dependencies so having all
software installed centrally in one loaction can cause conflicts. For
example if trying to use a software application that relies on python 2
and another software application that relies on python 3 it is
impossible to have them exist togther. Therefore we create software
environments (or boxes) that conatin only the software and depencies
needed and switch between them as needed. Now that we have activated the
kraken environment let's use it to look for contamination.

    \hypertarget{setting-up-a-database}{%
\subsection{Setting up a database}\label{setting-up-a-database}}

To run Kraken you need to either build a database or download an
existing one. The standard database is very large (33 GB), but
thankfully there are some smaller, pre-built databased available. To
download the smallest of them, the 4 GB MiniKraken, run:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
wget\PY{+w}{ }https://ccb.jhu.edu/software/kraken/dl/minikraken\PYZus{}20171019\PYZus{}4GB.tgz
\end{Verbatim}
\end{tcolorbox}

    This may take some time to run. So skip ahead to the next section on
Heterozygous SNPs and come back to this section when the download is
cmplete. (If you have trouble installing Kraken or downloading the
database then we have pre-generated the kraken report file.)

Onve the database is downloaded all you need to do is un-tar it:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
tar\PY{+w}{ }\PYZhy{}zxvf\PY{+w}{ }minikraken\PYZus{}20171019\PYZus{}4GB.tgz
\end{Verbatim}
\end{tcolorbox}

    This version of the database is constructed from complete bacterial,
archaeal, and viral genomes in RefSeq,

    \hypertarget{running-kraken}{%
\subsection{Running Kraken}\label{running-kraken}}

To run Kraken, you need to provide the path to the database you just
created. By default, the input files are assumed to be in FASTA format,
so in this case we also need to tell Kraken that our input files are in
FASTQ format, gzipped, and that they are paired end reads:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
kraken\PY{+w}{ }\PYZhy{}\PYZhy{}db\PY{+w}{ }./minikraken\PYZus{}20171013\PYZus{}4GB\PY{+w}{ }\PYZhy{}\PYZhy{}output\PY{+w}{ }kraken\PYZus{}results\PY{+w}{ }\PY{l+s+se}{\PYZbs{}}
\PY{+w}{    }\PYZhy{}\PYZhy{}fastq\PYZhy{}input\PY{+w}{ }\PYZhy{}\PYZhy{}gzip\PYZhy{}compressed\PY{+w}{ }\PYZhy{}\PYZhy{}paired\PY{+w}{ }\PY{l+s+se}{\PYZbs{}}
\PY{+w}{    }data/13681\PYZus{}1\PYZsh{}18\PYZus{}1.fastq.gz\PY{+w}{ }data/13681\PYZus{}1\PYZsh{}18\PYZus{}2.fastq.gz
\end{Verbatim}
\end{tcolorbox}

    The five columns in the file that's generated are:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  ``C''/``U'': one letter code indicating that the sequence was either
  classified or unclassified.
\item
  The sequence ID, obtained from the FASTA/FASTQ header.
\item
  The taxonomy ID Kraken used to label the sequence; this is 0 if the
  sequence is unclassified.
\item
  The length of the sequence in bp.
\item
  A space-delimited list indicating the LCA mapping of each k-mer in the
  sequence.
\end{enumerate}

To get a better overview you can create a kraken report:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
kraken\PYZhy{}report\PY{+w}{ }\PYZhy{}\PYZhy{}db\PY{+w}{ }./minikraken\PYZus{}20171013\PYZus{}4GB\PY{+w}{ }kraken\PYZus{}results\PY{+w}{ }\PYZgt{}\PY{+w}{ }kraken\PYZhy{}report
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{looking-at-the-results}{%
\subsection{Looking at the results}\label{looking-at-the-results}}

Let's have a closer look at the kraken\_report for the sample. If for
some reason your kraken-run failed there is a prebaked kraken-report at
data/kraken-report

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
head\PY{+w}{ }\PYZhy{}n\PY{+w}{ }\PY{l+m}{20}\PY{+w}{ }kraken\PYZhy{}report
\end{Verbatim}
\end{tcolorbox}

    The six columns in this file are:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Percentage of reads covered by the clade rooted at this taxon
\item
  Number of reads covered by the clade rooted at this taxon
\item
  Number of reads assigned directly to this taxon
\item
  A rank code, indicating (U)nclassified, (D)omain, (K)ingdom, (P)hylum,
  (C)lass, (O)rder, (F)amily, (G)enus, or (S)pecies. All other ranks are
  simply `-'.
\item
  NCBI taxonomy ID
\item
  Scientific name
\end{enumerate}

\hypertarget{exercises}{%
\subsection{Exercises}\label{exercises}}

\textbf{Q1: What is the most prevalent species in this sample?}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \textbf{Q2: Does this match the bactinspector results?}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \textbf{Q3: What percentage of reads could not be classified?}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \hypertarget{heterozygous-snps}{%
\subsection{Heterozygous SNPs}\label{heterozygous-snps}}

For bacteria, another thing that you can look at to detect contamination
is if there are heterozygous SNPs in your samples. Simply put, if you
align your reads to a reference, you would expect any SNPs to be
homozygous, i.e.~if one read differs from the reference genome, then the
rest of the reads that map to that same location will also do so:

\textbf{Homozygous SNP}\\
Ref:~~~~~~~CTTGAGACGAAATCACTAAAAAACGTGACGACTTG\\
Read1:~~CTTGAGtCG\\
Read2:~~CTTGAGtCGAAA\\
Read3:~~~~~~~~~GAGtCGAAATCACTAAAA\\
Read4:~~~~~~~~~~~~~~~GtCGAAATCA

But if there is contamination, this may not be the case. In the example
below, half of the mapped reads have the T allele and half have the A.

\textbf{Heterozygous SNP}\\
Ref:~~~~~~~CTTGAGACGAAATCACTAAAAAACGTGACGACTTG\\
Read1:~~CTTGAGtCG\\
Read2:~~CTTGAGaCGAAA\\
Read3:~~~~~~~~~GAGaCGAAATCACTAAAA\\
Read4:~~~~~~~~~~~~~~~GtCGAAATCA

One way to asses this is to map your reads to a reference genoe and acll
variants and use bcftools to count the number of heterzygous SNPs.

    \hypertarget{confindr}{%
\subsection{ConFindr}\label{confindr}}

An alternative to counting the number of heterozygous variants is to use
a tool called ConFindr. ConFindr is a pipeline that can detect
contamination in bacterial NGS data, both between and within species.
Instead of looking at SNPs/variants across the whole genome, ConFindr
works by looking at conserved core genes - either using rMLST genes (53
genes are known to be single copy and conserved across all bacteria with
some known exceptions, which ConFindr handles), or custom sets of genes
derived from core-genome schemes. As the genes ConFindr looks at are
single copy, any sample that has multiple alleles of one or more gene is
likely to be contaminated.

To read more information about ConFindr visit:

https://olc-bioinformatics.github.io/ConFindr/

Unfortunately we do not have time to run ConFindr here. But some of the
automatated high throughput pipelines for mapping and snp calling and
genome assembly include QC assment in the preprocessing steps like
FastQC, bactinspector and confindr. We will be covering these pipelines
later in this course.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
Congratulations!\PY{+w}{ }You\PY{+w}{ }have\PY{+w}{ }reached\PY{+w}{ }the\PY{+w}{ }end\PY{+w}{ }of\PY{+w}{ }this\PY{+w}{ }tutorial.
\end{Verbatim}
\end{tcolorbox}


    % Add a bibliography block to the postdoc



\end{document}
